### DISCUSSION TITLE
Where is all of the interpreter/debugger's state?

### DISCUSSION THREAD

**Discussion by ravenjoad** [cucapra/futil#2178] [Q&A]
I am tinkering with a project right now and need to find all for the component being interpreted/debugged. After reading through the `interp` library's code, and running through a few of the example tests, it seems that:
  1. Every component gets an interpreter, which contains either a control or structural interpreter.
  2. The only state an interpreter needs comes from the `InterpreterState` struct.

Is that assessment correct? 

Currently, the goal is to just write a trace of the interpreter state to disk, perhaps by extending the struct with a serde derivation.

**Comment by EclecticGriffin** [comment]
Hi there! I'm the Cider (interpreter/debugger) person, so feel free to direct all your questions or wrath in my direction.

Off the top it is worth mentioning that you are asking about the Cider 1 code which is **soon to be removed**---possibly by the end of the week if I code fast---in favor of the Cider 2 architecture (currently lives under the `flattened` directory in `interp`).

So to answer your questions: Your assessment is correct. I've offered a few elaborations on the given points.

> Every component gets an interpreter, which contains either a control or structural interpreter.

Yup! Every component _instance_ is one of these two depending on how it is defined. These interpreter objects implement the same trait interface as primitives, so components that own components are oblivious to whether their cells are primitives from the standard library or are themselves Calyx components.

> The only state an interpreter needs comes from the InterpreterState struct.

Yup, everything lives there in the `cell_map` and `port_map`. Everything else is basically bookkeeping. Worth noting that `InterpreterState` belongs implicitly to a single component instance with the control flow (or lack thereof) instantiated by the various "control" interpreters (or structural interpreters).


Can you tell me a bit more about what state you're hoping to trace?

**Reply by ravenjoad** [reply]
I am currently chiefly interested in dumping the state of a circuit to disk. I want to be able to snapshot the state of a circuit at a point in simulation/interpretation. I figure this will only have to include the stateful portions of the Calyx component, because purely structural (combinatorial) will be able to figure out their state as-needed using the surrounding state.

Since `InterpreterState` is the only thing holding all of that state, I figured I could slap a `#derive(...)` on the struct (and whatever sub-structs/things I need) and serialize to JSON or whatever for later viewing.

**Reply by EclecticGriffin** [reply]
Worth noting that there is already a `print_env` function which dumps all the ports and memory state for the environment. https://github.com/calyxir/calyx/blob/main/interp/src/structures/environment.rs#L561

This of course doesn't handle sub-components, since those are behind dyn boxes.

You could also generate a `vcd` trace through the calyx driver, though obviously there is some effort needed to parse names and values

**Reply by ravenjoad** [reply]
The end-goal is to re-load the interpreter state from the file dumped to disk. So I thought dumping to JSON using serde and re-loading it and setting the `InterpreterState` would be the way to do that. Since I want to reload the state for a circuit, I need those sub-components in the dumped format too.

**Reply by EclecticGriffin** [reply]
Ah, understood. So unfortunately that feature won't really be possible with the way Cider 1 works since ports and cells are keyed based on their raw pointers (i.e. memory location) from the IR which won't be deserializable. But the good news is that the Cider 2 redesign deliberately enables save/load mechanics. I've yet to implement that feature but I'll bump it up my list once the Cider 1 removal is finished.

In the meantime you can take a look at the program state for Cider 2 here: https://github.com/calyxir/calyx/blob/main/interp/src/flatten/structures/environment/env.rs#L218.

Unlike Cider 1, this `Environment` holds the _entire_ program state, not just a single component and everything is laid out using vectors keyed by stable indexes.

**Reply by rachitnigam** [reply]
@KarlJoad this seems like an interesting direction! Perhaps it would be useful to schedule a short meeting with @EclecticGriffin and others to get a sense of where Cider 2 is going and how it can be aligned to meet your needs as well/how you could contribute to the developing code base?

**Reply by EclecticGriffin** [reply]
Yup, happy to schedule a meeting if that would be helpful!

**Reply by ravenjoad** [reply]
Ah... Ok. I did not see/realize that the ports and cells in the Interpreter was reliant on the IR's raw pointer values in-memory. That would make the serdes step complicated, for sure.

I will definitely start looking at Cider 2, especially if you are releasing it in the next few weeks.

Definitely! I'm available for a meeting most days during the summer. Right now I am just toying with a project idea, so I am not 100% sure what my needs are quite yet.

**Reply by EclecticGriffin** [reply]
Cool! Just shoot me an email and we can get something on the calendar
